File: ml-service/tracing.py
import json
import time
from pathlib import Path
from typing import Sequence

from opentelemetry import trace
from opentelemetry.trace import StatusCode
from opentelemetry.sdk.trace import TracerProvider, ReadableSpan
from opentelemetry.sdk.trace.export import SimpleSpanProcessor, SpanExporter, SpanExportResult
from opentelemetry.sdk.resources import Resource
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.propagate import set_global_textmap
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

BASE_DIR = Path(__file__).resolve().parent
ROOT_DIR = BASE_DIR.parent
TRACE_LOG_PATH = ROOT_DIR / "traces.jsonl"


class FileSpanExporter(SpanExporter):
    """Exports spans directly to a local traces.jsonl file."""

    def __init__(self, file_path: Path = TRACE_LOG_PATH):
        self.file_path = file_path

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        try:
            records = []
            for span in spans:
                ctx = span.get_span_context()
                trace_id = format(ctx.trace_id, "032x")
                span_id = format(ctx.span_id, "016x")
                parent_span_id = format(span.parent.span_id, "016x") if span.parent else None

                duration_ns = (span.end_time - span.start_time) if (span.end_time and span.start_time) else 0
                duration_ms = round(duration_ns / 1e6, 2)

                status_str = "UNSET"
                if span.status.status_code == StatusCode.OK:
                    status_str = "OK"
                elif span.status.status_code == StatusCode.ERROR:
                    status_str = "ERROR"

                record = {
                    "service": "wealthgenie-ml-service",
                    "trace_id": trace_id,
                    "span_id": span_id,
                    "parent_span_id": parent_span_id,
                    "name": span.name,
                    "duration_ms": duration_ms,
                    "status": status_str,
                    "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(span.start_time / 1e9)) if span.start_time else time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                    "attributes": dict(span.attributes) if span.attributes else {},
                }
                records.append(json.dumps(record, default=str))

            if records:
                with open(self.file_path, "a", encoding="utf-8") as f:
                    f.write("\n".join(records) + "\n")
            return SpanExportResult.SUCCESS
        except Exception as e:
            print(f"[Tracing] Failed to export spans to file: {e}")
            return SpanExportResult.FAILURE

    def shutdown(self):
        pass


def setup_tracing(app) -> TracerProvider:
    """Configures global OpenTelemetry TracerProvider and instruments the FastAPI app."""
    # Set W3C TraceContext propagator globally
    set_global_textmap(TraceContextTextMapPropagator())

    resource = Resource.create({"service.name": "wealthgenie-ml-service"})
    provider = TracerProvider(resource=resource)
    exporter = FileSpanExporter()
    provider.add_span_processor(SimpleSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    FastAPIInstrumentor.instrument_app(app, tracer_provider=provider)
    return provider

File: server/config/tracing.js
import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { trace, propagation } from '@opentelemetry/api';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = path.resolve(__dirname, '../..');
const DEFAULT_TRACE_LOG_PATH = path.join(ROOT_DIR, 'traces.jsonl');
const SAFE_ATTRIBUTE_NAMES = new Set([
  'agent.type', 'agent.name', 'agent.version', 'agent.graph_version', 'agent.scaffold_version',
  'agent.run_id', 'agent.status', 'agent.step_count', 'agent.tool_call_count', 'agent.model_call_count',
  'agent.tool_name', 'agent.tool_outcome', 'agent.policy_result', 'agent.evidence_status',
  'agent.fallback_used', 'agent.candidate_id', 'agent.evaluation_partition', 'agent.evaluation_version',
  'gen_ai.system', 'gen_ai.request.model', 'gen_ai.response.model', 'gen_ai.operation.name',
  'gen_ai.response.finish_reasons', 'gen_ai.usage.input_tokens', 'gen_ai.usage.output_tokens',
  'gen_ai.usage.total_tokens', 'error.type',
]);
const SENSITIVE_ATTRIBUTE = /(user|profile|income|salary|tax|email|phone|address|prompt|content|input|output|secret|token|password|cookie|authorization|raw|payload|value)/i;

function safeAttributes(attributes = {}) {
  return Object.fromEntries(Object.entries(attributes).filter(([name, value]) => (
    SAFE_ATTRIBUTE_NAMES.has(name)
    && !SENSITIVE_ATTRIBUTE.test(name)
    && ['string', 'number', 'boolean'].includes(typeof value)
  )).map(([name, value]) => [name, typeof value === 'string' ? value.slice(0, 120) : value]));
}

export function resolveTraceLogPath(env = process.env) {
  return path.resolve(env.TRACE_LOG_PATH || DEFAULT_TRACE_LOG_PATH);
}

const TRACE_LOG_PATH = resolveTraceLogPath();

/**
 * Custom FileSpanExporter that appends OpenTelemetry spans to traces.jsonl
 */
export class FileSpanExporter {
  constructor(filePath = TRACE_LOG_PATH) {
    this.filePath = path.resolve(filePath);
    this.writeFailureReported = false;
  }

  export(spans, resultCallback) {
    try {
      const records = spans.map(span => {
        const traceId = span.spanContext().traceId;
        const spanId = span.spanContext().spanId;
        const parentSpanId = span.parentSpanId || null;
        const durationNano = span.duration[0] * 1e9 + span.duration[1];
        const durationMs = Number((durationNano / 1e6).toFixed(2));

        return JSON.stringify({
          service: 'wealthgenie-express',
          trace_id: traceId,
          span_id: spanId,
          parent_span_id: parentSpanId,
          name: span.name,
          duration_ms: durationMs,
          status: span.status.code === 1 ? 'OK' : (span.status.code === 2 ? 'ERROR' : 'UNSET'),
          timestamp: new Date(span.startTime[0] * 1000 + span.startTime[1] / 1e6).toISOString(),
          attributes: safeAttributes(span.attributes || {}),
        });
      });

      if (records.length > 0) {
        fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
        fs.appendFileSync(this.filePath, records.join('\n') + '\n', 'utf8');
      }
      this.writeFailureReported = false;
      resultCallback({ code: 0 }); // ExportResultCode.SUCCESS
    } catch (err) {
      if (!this.writeFailureReported) {
        console.warn('[Tracing] Failed to export spans to file:', err.message, `path=${this.filePath}`);
        this.writeFailureReported = true;
      }
      resultCallback({ code: 1, error: err }); // ExportResultCode.FAILED
    }
  }

  shutdown() {
    return Promise.resolve();
  }
}

const fileExporter = new FileSpanExporter();

const sdk = new NodeSDK({
  serviceName: 'wealthgenie-express',
  spanProcessor: new SimpleSpanProcessor(fileExporter),
  instrumentations: [
    getNodeAutoInstrumentations({
      '@opentelemetry/instrumentation-fs': { enabled: false },
      '@opentelemetry/instrumentation-dns': { enabled: false },
      '@opentelemetry/instrumentation-net': { enabled: false },
    }),
  ],
});

sdk.start();

export { trace, propagation, TRACE_LOG_PATH };
export default sdk;
